File: ch08/bucket_sort.cpp

#include <iostream>
#include <vector>
#include <random>
#include <functional>

using std::cout;
using std::vector;

template<typename BaseData>
void exchange(vector<BaseData>& v, int i, int j) {
    BaseData temp = v[i];
    v[i] = v[j];
    v[j] = temp;
}

template<typename BaseData>
void quick_sort(vector<BaseData>& v) {
    auto partition = [&](int l, int r) {
        int k = l - 1;
        for (int i = l; i != r; i += 1) {
            if (v[i] < v[r]) {
                k += 1;
                exchange(v, i, k);
            }
        }
        k += 1;
        exchange(v, k, r);
        return k;
    };

    std::function<void (int, int)> quick_sort_asc = [&](int l, int r) {
        if (l < r) {
            int q = partition(l, r);
            quick_sort_asc(l, q - 1);
            quick_sort_asc(q + 1, r);
        }
    };

    quick_sort_asc(0, v.size() - 1);
}

void bucket_sort(vector<double>& v) {
    int n = v.size();
    vector<double> bucket[n];

    for (int i = 0; i != n; i += 1) {
        int index = n * v[i];
        bucket[index].push_back(v[i]);
    }

    for (int i = 0; i != n; i += 1) {
        quick_sort(bucket[i]);
    }

    int k = 0;
    for (int i = 0; i != n; i += 1) {
        int size = (int)bucket[i].size();
        for (int j = 0; j != size; j += 1) {
            v[k] = bucket[i][j];
            k += 1;
        }
    }
}

//--------------------------------------

long random(long min, long max) {
    std::random_device rd;
    std::mt19937_64 gen(rd());
    std::uniform_int_distribution<long> dis(min, max - 1);
    return dis(gen);
}

double random(double min, double max) {
    std::random_device rd;
    std::mt19937_64 gen(rd());
    std::uniform_real_distribution<double> dis(min, max);
    return dis(gen);
}

void set_vector(vector<long>& v, int size, long min, long max) {
    for (int i = 0; i < size; i = i+1) {
        v.push_back(random(min, max));
    }
}

void set_vector(vector<double>& v, int size, double min, double max) {
    for (int i = 0; i < size; i += 1) {
        v.push_back(random(min, max));
    }
}

template<typename BaseData>
void print_vector(vector<BaseData>& v) {
    cout << "(vector #";
    cout << v.size() << " [";
    for (int i = 0; i < (int)v.size(); i = i+1) {
        cout << v[i] << " ";
    }
    cout << "])\n";
}

int main() {
    vector<double> A;
    set_vector(A, 113, 0, 1);
    print_vector(A);
    bucket_sort(A);
    print_vector(A);
    return 0;
}

File: ch08/bucket_sort.java

import java.util.ArrayList;
import java.util.Random;

class Sort {
  private static <T> void exchange(ArrayList<T> v, int i, int j) {
    T temp = v.get(i);
    v.set(i, v.get(j));
    v.set(j, temp);
  }

  private static void insertion(ArrayList<Double> v) {
    int n = v.size();
    if (n > 0) {
      for (int i = 1; i != n; i += 1) {
        for (int p = i - 1, q = i;
            p >= 0 && v.get(p) > v.get(q); p -= 1, q -= 1) {
          exchange(v, p, q);
        }
      }
    }
  }

  public static void bucket(ArrayList<Double> v) {

    int n = v.size();
    ArrayList<ArrayList<Double>> buk =
      new ArrayList<ArrayList<Double>>(n);
    for (int i = 0; i != n; i += 1) {
      buk.add(new ArrayList<>());
    }

    for (int i = 0; i != n; i += 1) {
      int index = (int)(n * v.get(i));
      buk.get(index).add(v.get(i));
    }

    for (int i = 0; i != n; i += 1) {
      insertion(buk.get(i));
    }

    int k = 0;
    for (int i = 0; i != n; i += 1) {
      int len = buk.get(i).size();
      for (int j = 0; j != len; j += 1) {
        v.set(k, buk.get(i).get(j));
        k += 1;
      }
    }
  }
}

//--------------------------------------

class Main {

  static void setRandomArray(
      ArrayList<Integer> v, int size, int min, int max) {
    Random rand = new Random();
    int bound = max - min;
    for (int i = 0; i < size; i += 1) {
      v.add(rand.nextInt(bound) + min);
    }
  }

  static void setRandomArray(
      ArrayList<Double> v, int size) {
    Random rand = new Random();
    for (int i = 0; i < size; i += 1) {
      v.add(rand.nextDouble());
    }
  }

  static <T> void printArray(ArrayList<T> v) {
    System.out.print("(ArrayList #");
    System.out.print(v.size() + " [");
    for (int i = 0; i < v.size(); i += 1) {
      System.out.print(v.get(i) + " ");
    }
    System.out.println("])");
  }

  public static void main(String[] args) {

    ArrayList<Double> A = new ArrayList<>();
    setRandomArray(A, 100);
    printArray(A);
    Sort.bucket(A);
    printArray(A);
  }
}
